authz/audit_logger.go

```diff
@@ -0,0 +1,111 @@
+	PolicyName string `json:"policy_name,omitempty"`
+	// MatchedRule is the matched rule (or policy name in the xDS RBAC filter). It will be
+	// empty if there is no match.
+	MatchedRule string `json:"matched_rule,omitempty"`
+	// Authorized indicates whether the audited RPC is authorized or not.
+	Authorized bool `json:"bool,omitempty"`
+}
+
+// AuditLoggerConfig defines the configuration for a particular implementation of audit logger.
+type AuditLoggerConfig interface {
+	// Name returns the same name as that returned by its supported builder.
+	Name() string
+	// auditLoggerConfig is a dummy interface requiring users to embed this
+	// interface to implement it.
+	auditLoggerConfig()
+}
+
+// AuditLogger is the interface for an audit logger.
+// An audit logger is a logger instance that can be configured to use via the authorization policy
+// or xDS HTTP RBAC filters. When the authorization decision meets the condition for audit, all the
+// configured audit loggers' Log() method will be invoked to log that event with the AuditInfo.
+// The method will be executed synchronously before the authorization is complete and the call is
+// denied or allowed.
+// Please refer to https://github.com/grpc/proposal/pull/346 for more details about audit logging.
+type AuditLogger interface {
+	// Log logs the auditing event with the given information.
+	Log(context.Context, *AuditInfo)
+}
+
+// AuditLoggerBuilder is the interface for an audit logger builder.
+// It parses and validates a config, and builds an audit logger from the parsed config. This enables
+// configuring and instantiating audit loggers in the runtime.
+// Users that want to implement their own audit logging logic should implement this along with
+// the AuditLogger interface and register this builder by calling RegisterAuditLoggerBuilder()
+// before they start the gRPC server.
+// Please refer to https://github.com/grpc/proposal/pull/346 for more details about audit logging.
+type AuditLoggerBuilder interface {
+	// ParseAuditLoggerConfig parses an implementation-specific config into a
+	// structured logger config this builder can use to build an audit logger.
+	// When users implement this method, its returned type must embed the
+	// AuditLoggerConfig interface.
+	ParseAuditLoggerConfig(config interface{}) (AuditLoggerConfig, error)
+	// Build builds an audit logger with the given logger config.
+	// This will only be called with valid configs returned from ParseAuditLoggerConfig()
+	// so implementers need to make sure it can return a logger without error
+	// at this stage.
+	Build(AuditLoggerConfig) AuditLogger
+	// Name returns the name of logger built by this builder.
+	// This is used to register and pick the builder.
+	Name() string
+}
+
+// GetAuditLoggerBuilder returns a builder with the given name.
+// It returns nil if the builder is not found in the registry.
+func GetAuditLoggerBuilder(name string) AuditLoggerBuilder {
```

Comment:
Nit: let's move this up next to the other registry-related functions.

Reply:
Done.